src/Servers/Kestrel/Core/src/Internal/Certificates/CertificateConfigLoader.cs

```diff
@@ -37,6 +37,9 @@ public CertificateConfigLoader(IHostEnvironment hostEnvironment, ILogger<Kestrel
         else if (certInfo.IsFileCert)
         {
             var certificatePath = Path.Combine(HostEnvironment.ContentRootPath, certInfo.Path!);
+            var fullChain = new X509Certificate2Collection();
+            fullChain.ImportFromPemFile(certificatePath);
+
             if (certInfo.KeyPath != null)
             {
                 var certificateKeyPath = Path.Combine(HostEnvironment.ContentRootPath, certInfo.KeyPath);
```

Comment:
In this if block, you could just make `certificate` be `fullChain[0]` (assuming it's non-empty), and then you could remove `certificate` from the collection; depending on what public API guarantees you're making.

The two halves of that sentence are:
* We've already done the work of loading the certificate instance, why go open the file and process the contents a second time?
* The perf of a chain build will be marginally faster if you remove unnecessary elements.  Since the target certificate is already specified as the target it won't need to be found from the collection.  By removing it early you save a "nope, not the one I'm looking for, next!"

If you don't remove it, because you want the "full" chain to be in the HttpsOptions.ServerCertificateChain collection, then you have to decide if you want to have the same instance in the property and the collection.  If "yes" to the full chain and "no" to the same instance... then leave the code as-is :smile:.

And the reason I put "full" in quotes is that I don't think Let's Encrypt puts their root cert in that file, so the "full" chain is really "the chain except the root".  But it's as "full" as that file is, I suppose.

Reply:
@adityamandaleeka @blowdart @davidfowl @halter73 I can't say I have anywhere near enough context for this feature request/area to be able to personally decide how this should work...

Do we want to just vote or something?  Personally I like what @bartonjs is suggesting, i.e. we just use fullChain[0] and remove the cert from the collection.

Reply:
Yea remove it. A chain is everything above the leaf in my mind. Need to document carefully though.

Reply:
I did a bunch of research here a year ago and I can dig up my notes. When you get the certs from certbot for lets encrypt you can get both the fullchain.pem or the chain.pem + cert.pem (https://community.letsencrypt.org/t/generating-cert-pem-chain-pem-and-fullchain-pem-from-order-certificate/78376/6). 

We should also support loading the full chai from PFX files if it is there. NGINX supports the full chain as well (https://serverfault.com/questions/987612/nginx-ssl-config-for-cert-pem-and-chain-pem).

I'd like us to support both providing just the intermediates without the leaf cert and the full chain and we can remove the leaf cert (assuming this is easy). That makes us a bit more friendly.

Reply:
I'm not sure I understand the nuances here.  I updated the PR to Jeremy's suggestion so we remove the cert from the chain if we have a key path.

Is pfx support something different or does that just work?

> I'd like us to support both providing just the intermediates without the leaf cert and the full chain and we can remove the leaf cert (assuming this is easy). That makes us a bit more friendly.

The PR currently only supports full chain right?  So is the intermediates only scenario we don't assume fullChain[0] is the leaf?

Reply:
There are some differences in using the cert from the chain vs the current GetCertificate method, as a bunch of tests now fail.  So I think I'm just leave this code alone for now, and I will file a new issue to track @davidfowl and @bartonjs suggestions which I'd imagine we can take in rc2 still.  I just want to try and get something in for rc1 this week

Reply:
@bartonjs so the MacOS error is: "Status: One or more certificates required to validate this certificate cannot be found."  This is a bit weird since this works fine on ubuntu and windows, but I know we've skipped on MacOS many of our other cert tests today... I'm tempted to just skip this on macOS and file an issue to look at this later, unless this is something that is indicative of something seriously being broken

I filed https://github.com/dotnet/aspnetcore/issues/43193 to track resolving the full chain/intermediate/removing a cert.

I think many tests were all failing because the new test I copied from runtime was nuking all of our test certs which broke all the other tests as part of its cleanup, I scoped its cleanup to only the certs it was creating and things seem happier now.